tcadmin/resources/resources.py

```diff
@@ -76,9 +76,9 @@ def evolve(self, **args):
         return attr.evolve(self, **args)
 
     def __str__(self):
-        rv = ["{t.underline}{id}{t.normal}:".format(t=t, id=self.id)]
+        rv = ["{id}:".format(id=self.id)]
         for a in attr.fields(self.__class__):
-            label = "  {t.bold}{a.name}{t.normal}:".format(t=t, a=a)
+            label = "  {a.name}:".format(a=a)
```

Comment:
I spent an hour tracking down why this was required :)

`blessings.Terminal()` decides, when it's created, whether it's attached to a console.  If not, it doesn't add any color or underline or anything like that.

So why does it behave differently in this PR and on the main branch of this repo?  I tried reverting this PR bit by bit, and it turns out the critical line is the import of `appconfig` in `tcadmin/tests/conftest.py`.  `tcadmin/appconfig.py` imports `tcadmin/resources/resources.py`, which sets up `blessings.Terminal`.

Pytest collects its tests with the stdout stream pointing to a fake file, which would cause `blessings.Terminal` to think it's not pointed to a console and do nothing.  But pytest runs `conftest.py` with stdout point to the screen, which causes `blessings.Terminal` to think it's attached to a console, and output text with underline and bold, causing the comparisons to fail.

Import order *shouldn't* matter this much, but it's being a real annoyance in this PR!

Here's the fix I set up locally:

```patch
commit cd6804fb04baa803a1847875f555eac04c3bb1df (HEAD -> pr135)
Author: Dustin J. Mitchell <dustin@mozilla.com>
Date:   Tue Aug 4 16:57:42 2020 +0000

    import appconfig inside a test fixture rather than at the top level of conftest.py

diff --git a/tcadmin/resources/resources.py b/tcadmin/resources/resources.py
index 40dfe0c..a1e9ce1 100644
--- a/tcadmin/resources/resources.py
+++ b/tcadmin/resources/resources.py
@@ -71,19 +71,19 @@ class Resource(object):
             self.kind, getattr(self, attr.fields(self.__class__)[0].name)
         )
 
     def evolve(self, **args):
         "Create a new resource like this one, but with the named attributes replaced"
         return attr.evolve(self, **args)
 
     def __str__(self):
-        rv = ["{id}:".format(id=self.id)]
+        rv = ["{t.underline}{id}{t.normal}:".format(t=t, id=self.id)]
         for a in attr.fields(self.__class__):
-            label = "  {a.name}:".format(a=a)
+            label = "  {t.bold}{a.name}{t.normal}:".format(t=t, a=a)
             formatted = a.metadata.get("formatter", lambda id, v: str(v))(
                 self.id, getattr(self, a.name)
             )
             if "\n" in formatted:
                 rv.append(label)
                 rv.append(textwrap.indent(formatted, "    "))
             else:
                 rv.append("{} {}".format(label, formatted))
diff --git a/tcadmin/tests/conftest.py b/tcadmin/tests/conftest.py
index b5160b8..475821a 100644
--- a/tcadmin/tests/conftest.py
+++ b/tcadmin/tests/conftest.py
@@ -1,16 +1,15 @@
 # -*- coding: utf-8 -*-
 
 # This Source Code Form is subject to the terms of the Mozilla Public License,
 # v. 2.0. If a copy of the MPL was not distributed with this file, You can
 # obtain one at http://mozilla.org/MPL/2.0/.
 
 import pytest
-from tcadmin.appconfig import AppConfig
 
 
 def pytest_addoption(parser):
     parser.addoption(
         "--skip-slow", action="store_true", default=False, help="skip slow tests"
     )
 
 
@@ -19,14 +18,19 @@ def pytest_collection_modifyitems(config, items):
         skip_slow = pytest.mark.skip(reason="skipping slow tests")
         for item in items:
             if "slow" in item.keywords:
                 item.add_marker(skip_slow)
 
 
 @pytest.fixture(scope="module")
 def appconfig():
+    # don't import this at the top level, as it results in `blessings.Terminal` being
+    # initialized in a situation where output is to a console, and it includes underlines
+    # and bold and colors in the output, causing test failures
+    from tcadmin.appconfig import AppConfig
+
     appconfig = AppConfig()
     appconfig.description_prefix = (
         "*DO NOT EDIT* - This resource is configured automatically.\n\n"
     )
     with AppConfig._as_current(appconfig):
         yield appconfig
```

Reply:
I never digged about blessings when fixing it, I just thought to remove it and it worked.
But found new information today, thanks to you.
Fixed it!